File: deob-ast/src/main/java/dev/openrs2/deob/ast/transform/BitMaskTransformer.java
package dev.openrs2.deob.ast.transform;

import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.expr.BinaryExpr;
import com.github.javaparser.ast.expr.IntegerLiteralExpr;
import com.google.common.collect.ImmutableSet;
import dev.openrs2.deob.ast.util.ExprUtils;

public final class BitMaskTransformer extends Transformer {
	private static final ImmutableSet<BinaryExpr.Operator> SHIFT_OPS = ImmutableSet.of(
		BinaryExpr.Operator.SIGNED_RIGHT_SHIFT,
		BinaryExpr.Operator.UNSIGNED_RIGHT_SHIFT
	);

	private static final ImmutableSet<BinaryExpr.Operator> BITWISE_OPS = ImmutableSet.of(
		BinaryExpr.Operator.BINARY_AND,
		BinaryExpr.Operator.BINARY_OR,
		BinaryExpr.Operator.XOR
	);

	@Override
	public void transform(CompilationUnit unit) {
		unit.findAll(BinaryExpr.class).forEach(expr -> {
			var shiftOp = expr.getOperator();
			var left = expr.getLeft();
			var shamtExpr = expr.getRight();

			if (!SHIFT_OPS.contains(shiftOp) || !left.isBinaryExpr() || !shamtExpr.isIntegerLiteralExpr()) {
				return;
			}

			var bitwiseExpr = left.asBinaryExpr();
			var bitwiseOp = bitwiseExpr.getOperator();
			var argExpr = bitwiseExpr.getLeft();
			var maskExpr = bitwiseExpr.getRight();

			if (!BITWISE_OPS.contains(bitwiseOp) || !ExprUtils.isIntegerOrLongLiteral(maskExpr)) {
				return;
			}

			var shamt = shamtExpr.asIntegerLiteralExpr().asInt();
			if (maskExpr.isIntegerLiteralExpr()) {
				var mask = maskExpr.asIntegerLiteralExpr().asInt();

				switch (shiftOp) {
				case SIGNED_RIGHT_SHIFT:
					mask >>= shamt;
					break;
				case UNSIGNED_RIGHT_SHIFT:
					mask >>>= shamt;
					break;
				default:
					throw new IllegalStateException();
				}

				maskExpr = new IntegerLiteralExpr(mask);
			} else {
				var mask = maskExpr.asLongLiteralExpr().asLong();

				switch (shiftOp) {
				case SIGNED_RIGHT_SHIFT:
					mask >>= shamt;
					break;
				case UNSIGNED_RIGHT_SHIFT:
					mask >>>= shamt;
					break;
				default:
					throw new IllegalStateException();
				}

				maskExpr = ExprUtils.createLong(mask);
			}

			expr.replace(new BinaryExpr(new BinaryExpr(argExpr.clone(), shamtExpr.clone(), shiftOp), maskExpr, bitwiseOp));
		});
	}
}

File: deob-ast/src/main/java/dev/openrs2/deob/ast/transform/BitMaskTransformer.kt
package dev.openrs2.deob.ast.transform

import com.github.javaparser.ast.CompilationUnit
import com.github.javaparser.ast.expr.BinaryExpr
import com.github.javaparser.ast.expr.IntegerLiteralExpr
import com.github.javaparser.ast.expr.LongLiteralExpr
import dev.openrs2.deob.ast.Library
import dev.openrs2.deob.ast.LibraryGroup
import dev.openrs2.deob.ast.util.checkedAsInt
import dev.openrs2.deob.ast.util.checkedAsLong
import dev.openrs2.deob.ast.util.toLongLiteralExpr
import dev.openrs2.deob.ast.util.walk
import javax.inject.Singleton

@Singleton
class BitMaskTransformer : Transformer() {
    override fun transformUnit(group: LibraryGroup, library: Library, unit: CompilationUnit) {
        unit.walk { expr: BinaryExpr ->
            val shiftOp = expr.operator
            val bitwiseExpr = expr.left
            val shamtExpr = expr.right

            if (shiftOp !in SHIFT_OPS || bitwiseExpr !is BinaryExpr || shamtExpr !is IntegerLiteralExpr) {
                return@walk
            }

            val bitwiseOp = bitwiseExpr.operator
            val argExpr = bitwiseExpr.left
            var maskExpr = bitwiseExpr.right

            if (bitwiseOp !in BITWISE_OPS) {
                return@walk
            }

            val shamt = shamtExpr.checkedAsInt()
            when (maskExpr) {
                is IntegerLiteralExpr -> {
                    var mask = maskExpr.checkedAsInt()

                    mask = when (shiftOp) {
                        BinaryExpr.Operator.SIGNED_RIGHT_SHIFT -> mask shr shamt
                        BinaryExpr.Operator.UNSIGNED_RIGHT_SHIFT -> mask ushr shamt
                        else -> error("Invalid shiftOp")
                    }

                    maskExpr = IntegerLiteralExpr(mask.toString())
                }
                is LongLiteralExpr -> {
                    var mask = maskExpr.checkedAsLong()

                    mask = when (shiftOp) {
                        BinaryExpr.Operator.SIGNED_RIGHT_SHIFT -> mask shr shamt
                        BinaryExpr.Operator.UNSIGNED_RIGHT_SHIFT -> mask ushr shamt
                        else -> error("Invalid shiftOp")
                    }

                    maskExpr = mask.toLongLiteralExpr()
                }
                else -> return@walk
            }

            expr.replace(BinaryExpr(BinaryExpr(argExpr.clone(), shamtExpr.clone(), shiftOp), maskExpr, bitwiseOp))
        }
    }

    private companion object {
        private val SHIFT_OPS = setOf(
            BinaryExpr.Operator.SIGNED_RIGHT_SHIFT,
            BinaryExpr.Operator.UNSIGNED_RIGHT_SHIFT
        )
        private val BITWISE_OPS = setOf(
            BinaryExpr.Operator.BINARY_AND,
            BinaryExpr.Operator.BINARY_OR,
            BinaryExpr.Operator.XOR
        )
    }
}
